core-rust/state-manager/src/protocol/protocol_state.rs

```diff
@@ -18,76 +19,173 @@ impl NodeProtocolUpdateExecutor {
-    pub fn resume_protocol_update_if_any(&self) {
-        match ProtocolUpdateProgress::resolve(self.database.lock().deref()) {
-            ProtocolUpdateProgress::UpdateInitiatedButNothingCommitted {
-                protocol_version_name,
-            } => {
-                info!(
-                    "Starting a {} protocol update execution",
-                    protocol_version_name
-                );
-                self.execute_protocol_update_actions(&protocol_version_name, 0);
-            }
-            ProtocolUpdateProgress::UpdateInProgress {
-                protocol_version_name,
-                last_batch_index: last_batch_idx,
-            } => {
-                let next_batch_idx = last_batch_idx.checked_add(1).unwrap();
-                info!(
-                    "Resuming a {} protocol update execution from batch idx {}",
-                    protocol_version_name, next_batch_idx
-                );
-                self.execute_protocol_update_actions(&protocol_version_name, next_batch_idx);
+    ///
+    /// Returns the resultant protocol version.
+    pub fn resume_protocol_update_if_any(&self) -> Option<ProtocolVersionName> {
+        let mut latest_enacted: Option<ProtocolVersionName> = None;
+
+        // We loop because we might need to run back-to-back protocol updates.
+        loop {
+            let progress = ProtocolUpdateProgress::resolve(self.database.lock().deref());
+            //=========================== Helpful Notice (to future me) ============================
+            // It may look sensible to inline `progress` below, but trust me, it isn't.
+            // You get a deadlock because rust doesn't drop the database lock guard as soon as you'd expect.
+            // This notice was installed after doing this refactor twice, 6 months apart...
+            // And then having to play the fun "hunt for the deadlock" game.
+            //======================================================================================
+            match progress {
+                ProtocolUpdateProgress::UpdateInitiatedButNothingCommitted {
+                    protocol_version_name,
+                } => {
+                    info!("[UPDATE: {protocol_version_name}] Starting protocol update");
+                    self.execute_protocol_update_actions(&protocol_version_name, 0, 0);
+                    latest_enacted = Some(protocol_version_name.clone());
+                }
+                ProtocolUpdateProgress::UpdateInProgress {
+                    protocol_version_name,
+                    last_batch_group_index,
+                    last_batch_index,
+                } => {
+                    info!("[UPDATE: {protocol_version_name}] Resuming protocol update");
+                    self.execute_protocol_update_actions(
+                        &protocol_version_name,
+                        last_batch_group_index,
+                        last_batch_index.checked_add(1).unwrap(),
```

Comment:
What happens if the `last_batch_index` is already the last?

Reply:
Good question! :)

So `last_batch_index` is not the very last batch in the update, because otherwise we will be matching on `ProtocolUpdateProgress::NotUpdating` or `ProtocolUpdateProgress::UpdateInitiatedButNothingCommitted` (in the immediate trigger case).

So the only edge case is that `last_batch_index` is the last in its batch group, but _not_ the final batch group.

In which case, this iterator is empty when we execute the protocol update `let remaining_batches = batches.into_iter().enumerate().skip(start_at_batch);`, and we immediate go onto the next batch group. The logging is maybe a tiny bit misleading, but the execution logic is sound.

I'll add a comment here to explain this 👍